feat(cms): revamp TW Reporter related articles search and display functionality

ericsen-tsai opened the pull request:
## Summary

This PR revamps a comprehensive feature for searching and displaying TW Reporter related articles in the Kids Reporter CMS. The implementation includes a new GraphQL query for searching TW Reporter posts, enhanced CMS interface for managing related articles.

## Changes

- **GraphQL Schema Enhancement**: Added `searchTWReporterPosts` query in `extend-schema.ts` to search TW Reporter articles using Google Custom Search API
- **CMS Interface Redesign**: 
  - Removed the old search interface (`search-related-posts.tsx`)
  - Enhanced `twreporter-related-posts.tsx` with integrated search functionality
  - Updated field label from '新增與排序' to '相關文章管理'
  - Added real-time search with GraphQL integration
- **Frontend Component Refactoring**:
  - Replaced `RelatedPosts` component with new `RelatedArticles` component
- **GraphQL Query Enhancement**: Added `TWReporterRelatedPostsJSON` field to post queries

## Additional Notes

- The search functionality uses Google Custom Search API with filtering for TW Reporter articles
- Articles are automatically filtered to show only those from twreporter.org domain
- The interface includes drag-and-drop reordering for related articles
- Search results show article metadata including title, description, image, and published date
- External links to TW Reporter articles include visual indicators and hover effects
- This PR is based on #801 

## Screenshot(s)


https://github.com/user-attachments/assets/2d2e72e4-bf99-46db-87f0-215bfb34f16a



## Reference(s)

N/A

ericsen-tsai committed:
feat(cms): revamp search functionality for TWReporter posts

- Added a new query `searchTWReporterPosts` to the GraphQL schema.
- Implemented the search logic in `extend-schema.ts` to fetch posts based on keywords.
- Removed the deprecated `searchTWReporterRelatedPosts` field from the Post type.
- Updated the `twreporter-related-posts` view to include a search input and display search results.
- Enhanced the user interface for managing related posts with search capabilities.

nickhsine reviewed (commented):
謝啦，這個 PR 修掉了 https://github.com/kids-reporter/kids-reporter-monorepo/issues/732 裡暴露 google programmable search key 的問題！

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -180,7 +180,58 @@ export const extendGraphqlSchema = graphql.extend(() => {
+        args: {
+          keywords: graphql.arg({ type: graphql.nonNull(graphql.String) }),
+        },
+        async resolve(root, args) {
```
thought:

目前 kids-cms.twreporter.org 還沒有開啟 IAP（需要確認一下為何沒有啟動），這邊是不是需要先加個 authentication validation，確認 context.session.data.role 是否合格？避免外部使用者直接打這個 query。

ericsen-tsai replied:
@nickhsine could you help me review this commit [fix(cms): add error handling for searchTWReporterPosts and add validation for unauthorized and forbidden](https://github.com/kids-reporter/kids-reporter-monorepo/pull/803/commits/a2bf5a04754182cdad458d39ba569c32d3002069)?

ericsen-tsai commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -180,7 +180,58 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              )
+              ?.map((item: any) => {
+                const metaTag = item?.pagemap?.metatags?.[0]
+                const publishedDate = new Date(
+                  item?.snippet
+                    ?.split('...')?.[0]
+                    .trim()
+                    .replace('年', '-')
+                    .replace('月', '-')
+                    .replace('日', '')
+                ).toISOString()
```
Comment added in [fix(frontend): modify style and add comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/801/commits/a8d7a63aa83818666036a950daffd7e3fce7df44)

nickhsine reviewed (approved):
LGTM 👍 👍 👍

ericsen-tsai commented on packages/cms/lists/views/twreporter-related-posts.tsx:
```diff
@@ -140,21 +169,87 @@ export const Field = ({
+            {relatedPosts.map((post: TWReporterPost, index: number) => {
               return (
-                <Draggable key={id} draggableId={id} index={index}>
+                <Draggable key={post.src} draggableId={post.src} index={index}>
```
All sources of posts are different

ericsen-tsai commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -352,6 +375,7 @@ const Article = ({ post }: { post: NonNullable<GetPostQuery['post']> }) => {
           <Divider />
           <div className="relative">
             <PostRenderer post={post} theme={theme} />
+            {/* middle of the article content enter 50% of the viewport*/}
```
Fixed in [fix(frontend): fix grammar errors](https://github.com/kids-reporter/kids-reporter-monorepo/pull/801/commits/8d1e0a24f7a6bfb0eec50185a016ef81eade187f)

ericsen-tsai commented on packages/frontend/src/app/(sticky-header)/_components/article/article.tsx:
```diff
@@ -387,8 +411,27 @@ const Article = ({ post }: { post: NonNullable<GetPostQuery['post']> }) => {
       </div>
+
+      <div className="relative w-full">
+        {/* related posts enter 50% of the viewport*/}
```
Fixed in [fix(frontend): fix grammar errors](https://github.com/kids-reporter/kids-reporter-monorepo/pull/801/commits/8d1e0a24f7a6bfb0eec50185a016ef81eade187f)

ericsen-tsai commented on packages/routing-ui/src/styles.css:
```diff
@@ -321,16 +207,16 @@
     font-weight: 700;
-    line-height: var(--line-height-tight);
-    letter-spacing: var(--letter-spacing-wide);
+    line-height: var(--line-height-normal);
+    letter-spacing: 1.6px;
```
Since other variables are used in multiple places, but letter-spacing isn't, I'll leave it as it is

ericsen-tsai commented on packages/routing-ui/src/styles.css:
```diff
@@ -321,16 +207,16 @@
     font-weight: 700;
-    line-height: var(--line-height-tight);
-    letter-spacing: var(--letter-spacing-wide);
+    line-height: var(--line-height-normal);
+    letter-spacing: 1.4px;
```
Since other variables are used in multiple places, but letter-spacing isn't, I'll leave it as it is

ericsen-tsai committed:
fix(cms): add error handling for searchTWReporterPosts and add validation for unauthorized and forbidden

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -186,8 +215,47 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                  errorCode,
+                },
+              })
+            )
```
thought:
因為這是 client error，severity 可能設定成 INFO 就好了？

ericsen-tsai replied:
Yes, you're right, it's not a system error.
But I think `INFO` is too low, so I used `WARNING` instead.

[fix(cms): improve error handling in generatePostQuestions and searchTWReporterPosts functions](https://github.com/kids-reporter/kids-reporter-monorepo/pull/803/commits/37c5056909dbff0c254af8cb7e477b345e0e3205)

Thanks!

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -224,10 +292,60 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                context: {
+                  function: 'searchTWReporterPosts',
+                  keywords,
+                },
```
nitpick:
好像可以把 data 移進 context 裡？
我之前好像也不小心寫在 context 外面。

ericsen-tsai replied:
Fixed in [fix(cms): improve error handling in generatePostQuestions and searchTWReporterPosts functions](https://github.com/kids-reporter/kids-reporter-monorepo/pull/803/commits/37c5056909dbff0c254af8cb7e477b345e0e3205)

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -224,10 +292,60 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                `Failed to search TW Reporter posts: ${_err.message}`,
+                {
+                  extensions: {
+                    code: 'EXTERNAL_SERVICE_ERROR',
```
question:
是不是統一用 `INTERNAL_SERVER_ERROR` 表達就好？
對於發送 request 到這個 resolver 的 client 來說， resolver 發生的非預期錯誤都算是 server error，
所以應該不用使用 axios 的 status code 來決定 http.status？

不過如果想要把 axios 的 error 印的完整一點，
可以考慮用 https://github.com/twreporter/twreporter-npm-packages/tree/master/packages/errors#errorshelpersannotateaxioserror 來 print axios error。

ericsen-tsai replied:
Sure, I already adjusted axios error with helper wrapper of twreporter errors

[fix(cms): improve error handling in generatePostQuestions and searchTWReporterPosts functions](https://github.com/kids-reporter/kids-reporter-monorepo/pull/803/commits/37c5056909dbff0c254af8cb7e477b345e0e3205)

ericsen-tsai committed:
fix(cms): improve error handling in generatePostQuestions and searchTWReporterPosts functions

- Added better error messages using @twreporter/errors for Axios errors.
- Updated logging to include response data in context for both functions.

nickhsine commented:
LGTM 💯 💯 
感謝幫忙優化 generatePostQuestion
